physics/sfc_diag.f

```diff
@@ -75,23 +95,131 @@ subroutine sfc_diag_run                                           &
 !       t2m(i)  = t2m(i) * sig2k
         wrk     = 1.0 - fhi
 
+        thcon    = (1.e5/ps(i))**rocp    
+       !-- make sure 1st level q is not higher than saturated value
+        qss    = fpvs(t1(i))
+        qss    = eps * qss / (ps(i) + epsm1 * qss)
+        q1c    = min(q1(i),qss) ! lev 1 spec. humidity
 
-        if(thsfc_loc) then ! Use local potential temperature
-          t2m(i)  = tskin(i)*wrk + t1(i)*prslki(i)*fhi - (grav+grav)/cp
-        else ! Use potential temperature referenced to 1000 hPa
-          t2m(i)  = tskin(i)*wrk + t1(i)*fhi - (grav+grav)/cp
-        endif
+        qv1    = q1c / (1. - q1c) ! lev 1 mixing ratio
+        qsfcmr   = qsurf(i)/(1. - qsurf(i)) ! surface mixing ratio
+        chs = cdq(i) * wind(i)
+        cqs = chs
+        qsfcprox = max(qmin,qv1 + evap(i)/cqs) ! surface mix. ratio computed from the flux
 
-        if(evap(i) >= 0.) then !  for evaporation>0, use inferred qsurf to deduce q2m
-          q2m(i) = qsurf(i)*wrk + max(qmin,q1(i))*fhi
-        else                   !  for dew formation, use saturated q at tskin
-          qss    = fpvs(tskin(i))
-          qss    = eps * qss / (ps(i) + epsm1 * qss)
-          q2m(i) = qss*wrk + max(qmin,q1(i))*fhi
-        endif
+        if(.not. flux) then
+        !-- original method
+          if(lsm /= lsm_ruc) then
+            if(thsfc_loc) then ! Use local potential temperature
+              t2m(i)=tskin(i)*wrk + t1(i)*prslki(i)*fhi - (grav+grav)/cp
+            else ! Use potential temperature referenced to 1000 hPa
+              t2m(i) = tskin(i)*wrk + t1(i)*fhi - (grav+grav)/cp
+            endif
+            if(evap(i) >= 0.) then !  for evaporation>0, use inferred qsurf to deduce q2m
+              q2m(i) = qsurf(i)*wrk + max(qmin,q1c)*fhi
+            else                   !  for dew formation, use saturated q at tskin
+              qss    = fpvs(tskin(i))
+              qss    = eps * qss/(ps(i) + epsm1 * qss)
+              q2m(i) = qss*wrk + max(qmin,q1c)*fhi
+            endif
+          else
+          !-- RUC lsm
+            if(chs.lt.1.E-5) then
+            !-- under very stable conditions use first level temperature
+              t2m(i) = t1(i)
+            else
+              t2m(i)  = tskin(i)*wrk + t1(i)*fhi - (grav+grav)/cp
+            endif
+        
+            if(cqs.lt.1.E-5) then
+              q2m(i)=max(qmin,q1c) ! spec. humidity
+            else
+              q2m(i) = qsurf(i)*wrk + max(qmin,q1c)*fhi
+            endif
+          endif ! RUC lsm
+
+        else
+        !-- flux method
+          if(chs2(i).lt.1.E-5) then
+          !-- under very stable conditions use first level temperature
+            t2m(i) = t1(i)
```

Comment:
same here. I thought this was proven to be a bad idea that resulted in speckling. This should be delete.